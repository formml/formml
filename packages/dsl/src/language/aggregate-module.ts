import { EmptyFileSystem, inject } from 'langium'
import {
  DefaultSharedModuleContext,
  LangiumSharedServices,
  createDefaultModule,
  createDefaultSharedModule,
} from 'langium/lsp'

import { FormMLModule, FormMLServices } from './formml/formml-module.js'
import { registerFormMLValidationChecks } from './formml/formml-validator.js'
import {
  FormMLDeclarationModule,
  FormMLDeclarationServices,
} from './formml-declaration/formml-declaration-module.js'
import {
  FormMLDeclarationGeneratedModule,
  FormMLGeneratedModule,
  FormMLGeneratedSharedModule,
} from './generated/module.js'
import LongestMatchServiceRegistry from './overrides/LongestMatchServiceRegistry.js'

/**
 * Create the full set of services required by Langium.
 *
 * First inject the shared services by merging two modules:
 *  - Langium default shared services
 *  - Services generated by langium-cli
 *
 * Then inject the language-specific services by merging three modules:
 *  - Langium default language-specific services
 *  - Services generated by langium-cli
 *  - Services specified in this file
 *
 * @param context Optional module context with the LSP connection
 * @returns An object wrapping the shared services and the language-specific services
 */
export function createAggregateServices(context: DefaultSharedModuleContext): {
  FormML: FormMLServices
  FormMLDeclaration: FormMLDeclarationServices
  shared: LangiumSharedServices
} {
  const shared = inject(
    createDefaultSharedModule(context),
    FormMLGeneratedSharedModule,
    {
      ServiceRegistry: () => new LongestMatchServiceRegistry(),
    },
  )
  const FormML = inject(
    createDefaultModule({ shared }),
    FormMLGeneratedModule,
    FormMLModule,
  )
  const FormMLDeclaration = inject(
    createDefaultModule({ shared }),
    FormMLDeclarationGeneratedModule,
    FormMLDeclarationModule,
  )
  shared.ServiceRegistry.register(FormML)
  shared.ServiceRegistry.register(FormMLDeclaration)

  registerFormMLValidationChecks(FormML)
  // registerFormMLDeclarationValidationChecks(FormMLDeclaration)

  if (!context.connection) {
    // We don't run inside a language server
    // Therefore, initialize the configuration provider instantly
    void shared.workspace.ConfigurationProvider.initialized({})
  }
  return { FormML, FormMLDeclaration, shared }
}

export const createInMemoryAggregateServices = () =>
  createAggregateServices(EmptyFileSystem)
