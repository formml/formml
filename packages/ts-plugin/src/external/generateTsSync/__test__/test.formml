form testForm {
  num  numField
  text textField
}
