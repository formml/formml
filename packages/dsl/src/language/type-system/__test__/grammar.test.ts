import { clearDocuments, parseHelper } from 'langium/test'

import type { FormMLDeclaration } from '../../generated/ast.js'

import { createInMemoryAggregateServices } from '../../aggregate-module.js'

describe('type system grammar', () => {
  const services = createInMemoryAggregateServices()

  afterEach(async () => {
    await clearDocuments(services.FormMLDeclaration)
  })

  const serialize = (ast: FormMLDeclaration) =>
    services.FormMLDeclaration.serializer.JsonSerializer.serialize(ast, {
      space: 2,
    })
  const parser = (input: string) =>
    parseHelper<FormMLDeclaration>(services.FormMLDeclaration)(input)
      .then((x) => x.parseResult)
      .then((r) =>
        r.lexerErrors.length > 0 || r.parserErrors.length > 0
          ? Promise.reject(
              new Error(
                'Parsing failed with errors:\n' +
                  [...r.lexerErrors, ...r.parserErrors]
                    .map((e) => e.message)
                    .join('\n'),
              ),
            )
          : r.value,
      )

  describe('type alias', () => {
    test('declare simple type alias', async () => {
      const input = `type MyType = text`
      const ast = await parser(input)
      expect(serialize(ast)).toMatchSnapshot()
    })

    test('type alias can be used as a type', async () => {
      const input = `
        type MyType = text
        fun myFunction(param: MyType): MyType
      `
      const ast = await parser(input)
      expect(serialize(ast)).toMatchSnapshot()
    })

    test('declare object type alias', async () => {
      const input = `type MyType = { name: text; age: num }`
      const ast = await parser(input)
      expect(serialize(ast)).toMatchSnapshot()
    })

    test('declare type alias with type parameters', async () => {
      const input = `type MyType<T, U> = { name: T; age: U }`
      const ast = await parser(input)
      expect(serialize(ast)).toMatchSnapshot()
    })

    test('call type alias with type parameters', async () => {
      const input = `
        type MyType<T, U> = { name: T; age: U }
        fun myFunction(param: MyType<text, num>): MyType<bool, decimal>
      `
      const ast = await parser(input)
      expect(serialize(ast)).toMatchSnapshot()
    })
  })
})
