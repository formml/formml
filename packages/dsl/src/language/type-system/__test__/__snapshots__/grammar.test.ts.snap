// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`type system grammar > type alias > declare object type alias 1`] = `
"{
  "$type": "FormMLDeclaration",
  "typeDeclarations": [
    {
      "$type": "TypeAliasDeclaration",
      "name": "MyType",
      "type": {
        "$type": "ObjectType",
        "properties": [
          {
            "$type": "Property",
            "name": "name",
            "type": {
              "$type": "TextType",
              "name": "text"
            }
          },
          {
            "$type": "Property",
            "name": "age",
            "type": {
              "$type": "NumType",
              "name": "num"
            }
          }
        ]
      },
      "typeParameters": []
    }
  ],
  "declarations": []
}"
`;

exports[`type system grammar > type alias > declare simple type alias 1`] = `
"{
  "$type": "FormMLDeclaration",
  "typeDeclarations": [
    {
      "$type": "TypeAliasDeclaration",
      "name": "MyType",
      "type": {
        "$type": "TextType",
        "name": "text"
      },
      "typeParameters": []
    }
  ],
  "declarations": []
}"
`;

exports[`type system grammar > type alias > declare type alias with type parameters 1`] = `
"{
  "$type": "FormMLDeclaration",
  "typeDeclarations": [
    {
      "$type": "TypeAliasDeclaration",
      "name": "MyType",
      "typeParameters": [
        {
          "$type": "TypeParameterDeclaration",
          "name": "T"
        },
        {
          "$type": "TypeParameterDeclaration",
          "name": "U"
        }
      ],
      "type": {
        "$type": "ObjectType",
        "properties": [
          {
            "$type": "Property",
            "name": "name",
            "type": {
              "$type": "TypeRef",
              "ref": {
                "$ref": "#/typeDeclarations@0/typeParameters@0"
              }
            }
          },
          {
            "$type": "Property",
            "name": "age",
            "type": {
              "$type": "TypeRef",
              "ref": {
                "$ref": "#/typeDeclarations@0/typeParameters@1"
              }
            }
          }
        ]
      }
    }
  ],
  "declarations": []
}"
`;

exports[`type system grammar > type alias > type alias can be used as a type 1`] = `
"{
  "$type": "FormMLDeclaration",
  "typeDeclarations": [
    {
      "$type": "TypeAliasDeclaration",
      "name": "MyType",
      "type": {
        "$type": "TextType",
        "name": "text"
      },
      "typeParameters": []
    }
  ],
  "declarations": [
    {
      "$type": "AnnotationDeclaration",
      "name": "myAnnotation",
      "parameters": [
        {
          "$type": "Parameter",
          "name": "param",
          "type": {
            "$type": "TypeRef",
            "ref": {
              "$ref": "#/typeDeclarations@0"
            }
          },
          "optional": false
        }
      ],
      "returnType": {
        "$type": "TypeRef",
        "ref": {
          "$ref": "#/typeDeclarations@0"
        }
      }
    }
  ]
}"
`;
